README
NAME
    Map::Tube::Sofia - Interface to the Sofia Metro Map.

SYNOPSIS
     use Map::Tube::Sofia;
     my $obj = Map::Tube::Sofia->new;
     my $routes_ar = $obj->get_all_routes($from, $to);
     my $lines_ar = $obj->get_lines;
     my $station = $obj->get_node_by_id($station_id);
     my $station = $obj->get_node_by_name($station_name);
     my $route = $obj->get_shortest_route($from, $to);
     my $stations_ar = $obj->get_stations($line);
     my $metro_name = $obj->name;
     my $xml_file = $obj->xml;

DESCRIPTION
    It currently provides functionality to find the shortest route between
    the two given nodes.

    For more information about Sofia Map, click here
    <https://en.wikipedia.org/wiki/Sofia_Metro>.

METHODS
    "new()"
             Constructor.

    "get_all_routes($from, $to)" [EXPERIMENTAL]
             Get all routes from station to station.
             Returns reference to array with Map::Tube::Route objects.

    "get_lines()"
             Get lines in metro map.
             Returns reference to array with Map::Tube::Line objects.

    "get_node_by_id($station_id)"
             Get station node by id.
             Returns Map::Tube::Node object.

    "get_node_by_name($station_name)"
             Get station node by name.
             Returns Map::Tube::Node object.

    "get_shortest_route($from, $to)"
             Get shortest route between $from and $to node names. Node names in $from and $to are case insensitive.
             Returns Map::Tube::Route object.

    "get_stations($line)"
             Get list of stations for concrete metro line.
             Returns reference to array with Map::Tube::Node objects.

    "name()"
             Get metro name.
             Returns string with metro name.

    "xml()"
             Get XML specification of Sofia metro.
             Returns string with XML.

EXAMPLE1
     # Pragmas.
     use strict;
     use warnings;

     # Modules.
     use Encode qw(decode_utf8 encode_utf8);
     use Map::Tube::Sofia;

     # Object.
     my $obj = Map::Tube::Sofia->new;

     # Get route.
     my $route = $obj->get_shortest_route(decode_utf8('Люлин'), decode_utf8('Вардар'));

     # Print out type.
     print "Route: ".encode_utf8($route)."\n";

     # Output:
     # Route: Люлин (Първи метродиаметър), Западен парк (Първи метродиаметър), Вардар (Първи метродиаметър)

EXAMPLE2
     # Pragmas.
     use strict;
     use warnings;

     # Modules.
     use Map::Tube::Sofia;

     # Object.
     my $obj = Map::Tube::Sofia->new;

     # Get XML file.
     my $xml_file = $obj->xml;

     # Print out XML file.
     print "XML file: $xml_file\n";

     # Output like:
     # XML file: .*/sofia-map.xml

EXAMPLE3
     # Pragmas.
     use strict;
     use warnings;

     # Modules.
     use Map::Tube::GraphViz;
     use Map::Tube::GraphViz::Utils qw(node_color_without_label);
     use Map::Tube::Sofia;

     # Object.
     my $obj = Map::Tube::Sofia->new;

     # GraphViz object.
     my $g = Map::Tube::GraphViz->new(
             'callback_node' => \&node_color_without_label,
             'driver' => 'neato',
             'tube' => $obj,
     ); 

     # Get graph to file.
     $g->graph('Sofia.png');

     # Print file.
     system "ls -l Sofia.png";

     # Output like:
     # -rw-r--r-- 1 skim skim 78098 Dec 23 11:47 Sofia.png

DEPENDENCIES
    File::Share, Map::Tube, Moo, namespace::clean.

SEE ALSO
    Map::Tube, Map::Tube::GraphViz, Map::Tube::Text::Table.

    Map::Tube::Barcelona, Map::Tube::Berlin, Map::Tube::Bucharest,
    Map::Tube::Delhi, Map::Tube::Dnipropetrovsk, Map::Tube::Kazan,
    Map::Tube::Kharkiv, Map::Tube::Kiev, Map::Tube::London,
    Map::Tube::Minsk, Map::Tube::Moscow, Map::Tube::Novosibirsk,
    Map::Tube::NYC, Map::Tube::Prague, Map::Tube::SaintPetersburg,
    Map::Tube::Samara, Map::Tube::Tbilisi, Map::Tube::Tokyo,
    Map::Tube::Vienna, Map::Tube::Warsaw, Map::Tube::Yekaterinburg.

REPOSITORY
    <https://github.com/tupinek/Map-Tube-Sofia>

AUTHOR
    Michal Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
     © 2014 Michal Špaček
     Artistic License
     BSD 2-Clause License

VERSION
    0.04



Changes
0.04 2014-12-23T11:57:18+01:00
 - Add EXAMPLE3 section to doc.
 - Add Map::Tube::GraphViz, Map::Tube::Text::Table,
   Map::Tube::Dnipropetrovsk, Map::Tube::Novosibirsk,
   Map::Tube::SaintPetersburg, Map::Tube::Vienna and
   Map::Tube::Yekaterinburg to SEE ALSO section in doc.
 - Add get_lines() method to doc.
 - Add get_stations() method to doc.
 - Remove executable bit from sofia-map.xml file.

0.03 2014-12-15T07:55:58+01:00
 - Add Map::Tube::Berlin and Map::Tube::Moscow to SEE ALSO section.
 - Add Map::Tube::Bucharest, Map::Tube::Kazan, Map::Tube::Kharkiv,
   Map::Tube::Samara and Map::Tube::Tbilisi to SEE ALSO section in doc.
 - Add Map::Tube::Kiev and Map::Tube::Minsk to SEE ALSO section.
 - Add get_all_routes(), get_node_by_id() and get_node_by_name() methods to
   doc.
 - Add metro name.
 - Describe get_all_routes() method as EXPERIMENTAL.
 - Fix REPOSITORY section in doc.
 - Fix get_shortest_route() method description in doc. Thanks Mohammad S
   Anwar (#100768).
 - Improve abstract to readable version.

0.02 2014-12-04T07:52:59+01:00
 - Add Map::Tube::Warsaw to SEE ALSO section in doc.
 - Add tests for get_shortest_route() method.
 - Fix XML map (Thanks Mohammad S Anwar). 
 - Remove obsolete pragma from EXAMPLE2 section in doc.
 - Update LICENSE AND COPYRIGHT section in doc.
 - Update examples/ directory.

0.01 2014-12-01T08:57:30+01:00
 - First version


Makefile.PL
# Pragmas.
use strict;
use warnings;

# Modules.
use inc::Module::Install;

# Definition.
abstract 'Interface to the Sofia Metro Map.';
author 'Michal Spacek <skim@cpan.org>';
install_share 'share';
license 'bsd';
license 'artistic_2';
name 'Map-Tube-Sofia';
readme_from 'Sofia.pm';
requires 'File::Share' => 0;
requires 'Map::Tube' => '2.41';
requires 'Moo' => 0;
requires 'namespace::clean' => 0;
requires 'perl' => '5.8.0';
resources 'bugtracker' => 'https://rt.cpan.org/Public/Dist/Display.html'.
	'?Name=Map-Tube-Sofia';
resources 'homepage' => 'https://github.com/tupinek/Map-Tube-Sofia';
resources 'repository' => 'git://github.com/tupinek/Map-Tube-Sofia';
test_requires 'English' => 0;
test_requires 'File::Object' => 0.08;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
test_requires 'Test::Pod' => 0;
test_requires 'Test::Pod::Coverage' => 0;
tests_recursive;
version '0.04';

# Run.
WriteAll();


Sofia.pm
package Map::Tube::Sofia;

# Pragmas.
use strict;
use warnings;
use 5.006;

# Modules.
use File::Share ':all';
use Moo;
use namespace::clean;

# Version.
our $VERSION = 0.04;

# Get XML.
has xml => (
	'is' => 'ro',
	'default' => sub {
		return dist_file('Map-Tube-Sofia', 'sofia-map.xml');
	},
);

with 'Map::Tube';

1;

__END__

=encoding utf8

=head1 NAME

Map::Tube::Sofia - Interface to the Sofia Metro Map.

=head1 SYNOPSIS

 use Map::Tube::Sofia;
 my $obj = Map::Tube::Sofia->new;
 my $routes_ar = $obj->get_all_routes($from, $to);
 my $lines_ar = $obj->get_lines;
 my $station = $obj->get_node_by_id($station_id);
 my $station = $obj->get_node_by_name($station_name);
 my $route = $obj->get_shortest_route($from, $to);
 my $stations_ar = $obj->get_stations($line);
 my $metro_name = $obj->name;
 my $xml_file = $obj->xml;

=head1 DESCRIPTION

It currently provides functionality to find the shortest route between the two
given nodes.

For more information about Sofia Map, click L<here|https://en.wikipedia.org/wiki/Sofia_Metro>.

=head1 METHODS

=over 8

=item C<new()>

 Constructor.

=item C<get_all_routes($from, $to)> [EXPERIMENTAL]

 Get all routes from station to station.
 Returns reference to array with Map::Tube::Route objects.

=item C<get_lines()>

 Get lines in metro map.
 Returns reference to array with Map::Tube::Line objects.

=item C<get_node_by_id($station_id)>

 Get station node by id.
 Returns Map::Tube::Node object.

=item C<get_node_by_name($station_name)>

 Get station node by name.
 Returns Map::Tube::Node object.

=item C<get_shortest_route($from, $to)>

 Get shortest route between $from and $to node names. Node names in $from and $to are case insensitive.
 Returns Map::Tube::Route object.

=item C<get_stations($line)>

 Get list of stations for concrete metro line.
 Returns reference to array with Map::Tube::Node objects.

=item C<name()>

 Get metro name.
 Returns string with metro name.

=item C<xml()>

 Get XML specification of Sofia metro.
 Returns string with XML.

=back

=head1 EXAMPLE1

 # Pragmas.
 use strict;
 use warnings;

 # Modules.
 use Encode qw(decode_utf8 encode_utf8);
 use Map::Tube::Sofia;

 # Object.
 my $obj = Map::Tube::Sofia->new;

 # Get route.
 my $route = $obj->get_shortest_route(decode_utf8('Люлин'), decode_utf8('Вардар'));

 # Print out type.
 print "Route: ".encode_utf8($route)."\n";

 # Output:
 # Route: Люлин (Първи метродиаметър), Западен парк (Първи метродиаметър), Вардар (Първи метродиаметър)

=head1 EXAMPLE2

 # Pragmas.
 use strict;
 use warnings;

 # Modules.
 use Map::Tube::Sofia;

 # Object.
 my $obj = Map::Tube::Sofia->new;

 # Get XML file.
 my $xml_file = $obj->xml;

 # Print out XML file.
 print "XML file: $xml_file\n";

 # Output like:
 # XML file: .*/sofia-map.xml

=head1 EXAMPLE3

 # Pragmas.
 use strict;
 use warnings;

 # Modules.
 use Map::Tube::GraphViz;
 use Map::Tube::GraphViz::Utils qw(node_color_without_label);
 use Map::Tube::Sofia;

 # Object.
 my $obj = Map::Tube::Sofia->new;

 # GraphViz object.
 my $g = Map::Tube::GraphViz->new(
         'callback_node' => \&node_color_without_label,
         'driver' => 'neato',
         'tube' => $obj,
 ); 

 # Get graph to file.
 $g->graph('Sofia.png');

 # Print file.
 system "ls -l Sofia.png";

 # Output like:
 # -rw-r--r-- 1 skim skim 78098 Dec 23 11:47 Sofia.png

=begin html

<a href="https://raw.githubusercontent.com/tupinek/Map-Tube-Sofia/master/images/Sofia.png">
  <img src="https://raw.githubusercontent.com/tupinek/Map-Tube-Sofia/master/images/Sofia.png" alt="Софийско метро" width="300px" height="300px" />
</a>

=end html

=head1 DEPENDENCIES

L<File::Share>,
L<Map::Tube>,
L<Moo>,
L<namespace::clean>.

=head1 SEE ALSO

L<Map::Tube>,
L<Map::Tube::GraphViz>,
L<Map::Tube::Text::Table>.

L<Map::Tube::Barcelona>,
L<Map::Tube::Berlin>,
L<Map::Tube::Bucharest>,
L<Map::Tube::Delhi>,
L<Map::Tube::Dnipropetrovsk>,
L<Map::Tube::Kazan>,
L<Map::Tube::Kharkiv>,
L<Map::Tube::Kiev>,
L<Map::Tube::London>,
L<Map::Tube::Minsk>,
L<Map::Tube::Moscow>,
L<Map::Tube::Novosibirsk>,
L<Map::Tube::NYC>,
L<Map::Tube::Prague>,
L<Map::Tube::SaintPetersburg>,
L<Map::Tube::Samara>,
L<Map::Tube::Tbilisi>,
L<Map::Tube::Tokyo>,
L<Map::Tube::Vienna>,
L<Map::Tube::Warsaw>,
L<Map::Tube::Yekaterinburg>.

=head1 REPOSITORY

L<https://github.com/tupinek/Map-Tube-Sofia>

=head1 AUTHOR

Michal Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

 © 2014 Michal Špaček
 Artistic License
 BSD 2-Clause License

=head1 VERSION

0.04

=cut
